### PR TITLE
Fix EXEC / DISCARD without MULTI wrongly reset watch

### PR DESCRIPTION
**Pull request by enjoy-binbin**
The old code call ResetWatchedKeys before the error, and then
ResetWatchedKeys reset watched_keys_modified flag which cause
the following inconsistencies (with Redis).

kvrocks output:
```
127.0.0.1:6666> watch a
OK
127.0.0.1:6666> watch b     ->>> other client touch a
OK
127.0.0.1:6666> exec
(error) ERR EXEC without MULTI
127.0.0.1:6666> multi
OK
127.0.0.1:6666(TX)> ping
QUEUED
127.0.0.1:6666(TX)> exec
1) PONG
```

redis output:
```
127.0.0.1:6379> watch a
OK
127.0.0.1:6379> watch b     ->>> other client touch a
OK
127.0.0.1:6379> exec
(error) ERR EXEC without MULTI
127.0.0.1:6379> multi
OK
127.0.0.1:6379(TX)> ping
QUEUED
127.0.0.1:6379(TX)> exec
(nil)
````

Introduced in #1279

### CODE REVIEW

**Review by git-hulk — DISMISSED** [review]

**Comment by enjoy-binbin** [comment]
ok now i see DISCARD also have the same problem, should I fix it together in this PR, or in a new PR?

**Comment by git-hulk** [comment]
> ok now i see DISCARD also have the same problem, should I fix it together in this PR, or in a new PR?

It’s fine to append a new commit.☺️

File: src/commands/cmd_txn.cc [hunk 1]
```diff
@@ -45,14 +45,14 @@ class CommandMulti : public Commander {
...
   Status Execute(Server *svr, Connection *conn, std::string *output) override {
-    auto reset_watch = MakeScopeExit([svr, conn] { svr->ResetWatchedKeys(conn); });
-
     if (!conn->IsFlagEnabled(Connection::kMultiExec)) {
       *output = redis::Error("ERR DISCARD without MULTI");
       return Status::OK();
     }
 
+    auto reset_watch = MakeScopeExit([svr, conn] { svr->ResetWatchedKeys(conn); });
     conn->ResetMultiExec();
```
Comment (enjoy-binbin) [inline src/commands/cmd_txn.cc:54]: Not sure if i should change it like this (consistent with EXEC)
```suggestion
    auto reset_multiexec = MakeScopeExit([conn] { conn->ResetMultiExec(); });
```

Comment (PragmaTwice) [reply src/commands/cmd_txn.cc:54]: Just keeping the original form is fine here.

**Review by git-hulk — APPROVED** [review]

**Review by mapleFU — APPROVED** [review]

**Comment by enjoy-binbin** [comment]
i see there is a flaky test:
https://github.com/apache/kvrocks/actions/runs/5495230368/jobs/10014556657?pr=1562#step:12:114
```
--- FAIL: TestString (46.13s)
    server.go:109: 
        	Error Trace:	/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:109
        	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:112
        	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:101
        	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/strings_test.go:742
        	Error:      	An error is expected but got nil.
        	Test:       	TestString
FAIL
exit status 1
```

But from the error message, it seems that it is hard to see which case failed (or which line failed). or i am searching it wrong?

**Pull request merged**